File: api/backend/schemas/maintenance.py
from __future__ import annotations

import uuid
from datetime import datetime
from typing import Optional

from pydantic import BaseModel, ConfigDict, model_validator


class MaintenanceWindowCreate(BaseModel):
    name: str
    description: Optional[str] = None
    device_selector: Optional[dict] = None  # None = all devices in tenant
    starts_at: datetime
    ends_at: datetime
    is_recurring: bool = False
    recurrence_cron: Optional[str] = None

    @model_validator(mode="after")
    def check_times(self) -> "MaintenanceWindowCreate":
        if self.ends_at <= self.starts_at:
            raise ValueError("ends_at must be after starts_at")
        if self.is_recurring and not self.recurrence_cron:
            raise ValueError("recurrence_cron is required when is_recurring is true")
        return self


class MaintenanceWindowUpdate(BaseModel):
    name: Optional[str] = None
    description: Optional[str] = None
    device_selector: Optional[dict] = None
    starts_at: Optional[datetime] = None
    ends_at: Optional[datetime] = None
    is_recurring: Optional[bool] = None
    recurrence_cron: Optional[str] = None


class MaintenanceWindowRead(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    tenant_id: uuid.UUID
    name: str
    description: Optional[str] = None
    device_selector: Optional[dict] = None
    starts_at: datetime
    ends_at: datetime
    is_recurring: bool
    recurrence_cron: Optional[str] = None
    created_by: Optional[uuid.UUID] = None
    created_at: datetime
    updated_at: datetime
    is_active: bool = False  # populated by the API, not stored

File: frontend/dashboard/src/api/maintenance.ts
import api from './client'

export interface MaintenanceWindow {
  id: string
  tenant_id: string
  name: string
  description: string | null
  device_selector: Record<string, unknown> | null
  starts_at: string
  ends_at: string
  is_recurring: boolean
  recurrence_cron: string | null
  created_by: string | null
  created_at: string
  updated_at: string
  is_active: boolean
  next_fire_at: string | null
}

export interface MaintenanceWindowCreate {
  name: string
  description?: string
  device_selector?: Record<string, unknown> | null
  starts_at: string
  ends_at: string
  is_recurring: boolean
  recurrence_cron?: string | null
}

export const fetchMaintenanceWindows = (params?: { device_id?: string; active_only?: boolean }) =>
  api.get<MaintenanceWindow[]>('/maintenance-windows', { params }).then(r => r.data)

export const createMaintenanceWindow = (body: MaintenanceWindowCreate) =>
  api.post<MaintenanceWindow>('/maintenance-windows', body).then(r => r.data)

export const deleteMaintenanceWindow = (id: string) =>
  api.delete(`/maintenance-windows/${id}`)
